packages/firestore/src/remote/remote_store.ts

```diff
@@ -55,6 +55,7 @@ import {
 import { ByteString } from '../util/byte_string';
 import { isIndexedDbTransactionError } from '../local/simple_db';
 import { User } from '../auth/user';
+import { rejectListen, applyRemoteEvent } from '../core/sync_engine';
```

Comment:
This creates a strong cycle between the SyncEngine and the RemoteStore (and also breaks our ability to inject an alternate implementation for testing, though I don't think we've ever done that).

Could you leave the RemoteSyncer interface as having those two methods, and then make function pointers of those names members of `SyncEngine` so that it still implements the interface, just not via its prototype?

Reply:
I thought long about this (especially since this change made the mocking in the EventListener test much more complex). I don' think there is a way to achieve the code savings that I am after if we use one shared interface. 

Instead, we would have to split RemoteSyncer into at least two interfaces (one for write handling, one for watch). On top of that, we would also need to inject the actual implementation at runtime, similar to what we are now doing in SyncEngine. If we provided the implementation at instantiation, the code would be included no matter what.

Reply:
Don't we have to solve the cycle problem though? Given all the trouble we've been having with random reorderings breaking things it seems unwise for us to intentionally create such cycles.

Meanwhile, what I'm proposing is that we have `RemoteStore` have a `RemoteSyncer` member whose members are fields that can be assigned. Then in the syncEngine once we get to a place where those callbacks are going to be required in the RemoteStore we assign to the fields. This gives essentially the same split you get here while preserving the flexibility in tests.